tethysext/atcore/models/resource_workflow_results/plot_workflow_result.py

```diff
@@ -0,0 +1,127 @@
+        'polymorphic_identity': TYPE
+    }
+
+    def __init__(self, *args, **kwargs):
+        """
+        Constructor.
+
+        Args:
+        """
+        super().__init__(*args, **kwargs)
+
+    @property
+    def default_options(self):
+        """
+        plot_lib (str): bokeh or plotly
+        axes(list): A list of tuples for pair axis ex. For example: [('x', 'y'), ('x1', 'y1'), ('x', 'y2')]
+        labels(list): A list of series' label. For example: ['Series 1', 'Series 2', 'Series 3']
+        axis_labels(list): A list of label for x and y axes respectively. For example: ['x', 'y']
+        plot_type (str): lines or scatter
+        line_shape (str): Only for plotly. You can select from on of these options: linear, spline, vhv, hvh, vh, hv
+
+        Returns default options dictionary for the object.
+        """
+        default_options = super().default_options
+
+        default_options.update({
+            'plot_lib': 'plotly',
+            'axes': [],
+            'labels': [],
+            'plot_type': 'lines',
+            'axis_labels': ['x', 'y'],
+            'line_shape': 'linear',
+            'no_dataset_message': 'No dataset found.'
+        })
+        return default_options
+
+    @property
+    def datasets(self):
+        if 'datasets' not in self.data:
+            self.data['datasets'] = []
+        return copy.deepcopy(self.data['datasets'])
+
+    @datasets.setter
+    def datasets(self, value):
+        data = copy.deepcopy(self.data)
+        data['datasets'] = value
+        self.data = data
+
+    def reset(self):
+        self.datasets = []
+
+    def _add_dataset(self, dataset):
+        """
+        Adds the dataset to the datasets array.
+
+        Args:
+            dataset(dict): The data.
+        """
+        datasets = self.datasets
+        datasets.append(dataset)
+        self.datasets = datasets
+
+    def add_pandas_dataframe(self, title, data_frame):
+        """
+        Adds a pandas.DataFrame to the result.
+
+        Args:
+            title(str): Display name.
+            data_frame(pandas.DataFrame): The data.
+        """
+
+        if not title:
+            raise ValueError('The argument "title" is required.')
+
+        if not isinstance(data_frame, pd.DataFrame):
+            raise ValueError('The argument "data_frame" must be a pandas.DataFrame.')
+
+        if data_frame.empty:
+            raise ValueError('The pandas.DataFrame must not be empty.')
+
+        d = {
+            'title': title,
+            'dataset': data_frame,
+        }
+        self._add_dataset(d)
+
+    def add_plot(self, plot):
+        """
+        Adds a pandas.DataFrame to the result.
+
+        Args:
+            plot(obj): bokeh or plotly figure.
+        """
+
+        if not isinstance(plot, go.Figure):
+            raise ValueError('The argument "plot" must be a plotly Figure.')
+
+        d = {
+            'plot_object': plot,
+        }
+
+        self._add_dataset(d)
```

Comment:
Create a new custom property in `self.data` for the plot object instead of storing it in the `self.datasets` list. Perhaps a `self.figure` property that assigns self.data['figure'] underneath? The goal is to hide the interaction with `self.data` as much as possible.